sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/PublishResultHandler.java

```diff
@@ -0,0 +1,90 @@
+ */
+package org.apache.beam.sdk.io.solace.broker;
+
+import com.solacesystems.jcsmp.JCSMPException;
+import com.solacesystems.jcsmp.JCSMPStreamingPublishCorrelatingEventHandler;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import org.apache.beam.sdk.io.solace.data.Solace;
+import org.apache.beam.sdk.io.solace.data.Solace.PublishResult;
+import org.apache.beam.sdk.io.solace.write.PublishResultsReceiver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is required to handle callbacks from Solace, to find out if messages were actually
+ * published or there were any kind of error.
+ *
+ * <p>This class is also used to calculate the latency of the publication. The correlation key
+ * contains the original timestamp of when the message was sent from the pipeline to Solace. The
+ * comparison of that value with the clock now, using a monotonic clock, is understood as the
+ * latency of the publication
+ */
+public final class PublishResultHandler implements JCSMPStreamingPublishCorrelatingEventHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PublishResultHandler.class);
+
+  @Override
+  public void handleErrorEx(Object key, JCSMPException cause, long timestamp) {
+    processKey(key, false, Optional.ofNullable(cause));
+  }
+
+  @Override
+  public void responseReceivedEx(Object key) {
+    processKey(key, true, Optional.empty());
+  }
+
+  private void processKey(Object key, boolean isPublished, Optional<JCSMPException> cause) {
+    PublishResult.Builder resultBuilder = PublishResult.builder();
+    String messageId;
+    if (key == null) {
+      messageId = "NULL";
+    } else if (key instanceof Solace.CorrelationKey) {
+      messageId = ((Solace.CorrelationKey) key).getMessageId();
+      long latencyMillis = calculateLatency((Solace.CorrelationKey) key);
+      resultBuilder = resultBuilder.setLatencyMilliseconds(latencyMillis);
+    } else {
+      messageId = key.toString();
+    }
+
+    resultBuilder = resultBuilder.setMessageId(messageId).setPublished(isPublished);
+    if (!isPublished) {
+      if (cause.isPresent()) {
+        resultBuilder = resultBuilder.setError(cause.get().getMessage());
+      } else {
+        resultBuilder = resultBuilder.setError("NULL - Not set by Solace");
+      }
+    } else if (cause.isPresent()) {
+      LOG.warn(
+          "Message with id {} is published but exception is populated. Ignoring exception",
+          messageId);
+    }
+
+    PublishResult publishResult = resultBuilder.build();
+    // Static reference, it receives all callbacks from all publications
+    // from all threads
+    PublishResultsReceiver.addResult(publishResult);
+  }
+
+  private static long calculateLatency(Solace.CorrelationKey key) {
+    long currentMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
+    long publishMillis = key.getPublishMonotonicMillis();
+    return currentMillis - publishMillis;
+  }
```

Comment:
Why are these nanos truncated to millis? It takes up the same amount of space in the correlation key anyhow.
If it's a presentational thing, you should probably deal with that when the metric is reported, but not before it's calculated.

Reply:
Done in upcoming commit.